File: perceptron.cpp
#include "perceptron.h"

Perceptron * makePerceptron(int len) {
    Perceptron * p = new Perceptron;
    p->len = len;
    p->b = 0.0;
    p->w = new float[len];
    p->a = 0.01;

    int i;
    for (i = 0; i < len; i++) {
        p->w[i] = 0.0;
    }

    return p;
}

float y(Perceptron * p, int * i) {
    float dot = 0.0;

    int j;
    for (j = 0; j < p->len; j++) {
        dot += p->w[j] * i[j];
    }

    dot += p->b;

    return dot > 0.0;
}

bool f(Perceptron * p, int * i) {
    return y(p, i) > 0.0;
}

void update(Perceptron * p, int * i, bool out) {
    int d = out ? 1 : 0;
    float yi = y(p, i);

    int j;
    for (j = 0; j < p->len; j++) {
        p->w[j] += (p->a) * (i[j]) * (d - yi);
    }

    p->b += (p->a) * (d - yi);
}

File: Perceptronus/src/com/gilljanssen/perceptronus/Perceptron.java
package com.gilljanssen.perceptronus;

public class Perceptron {

    // the learning rate of the perceptron
    final float rate;
    // the length of the weight vector
    final int len;
    // the bias factor
    float bias;
    // the weight vector
    final float[] weights;

    public Perceptron(int len) {
        this.rate = 0.01f;
        this.bias = 0.0f;
        this.len = len;
        this.weights = new float[len];

        for (float weight : weights) {
            weight = 0.0f;
        }
    }

    public float dotProduct(int[] features) {
        float dotp = bias;

        for (int i = 0; i < len; i++) {
            dotp += weights[i] * features[i];
        }

        return dotp;
    }

    public boolean fireNeuron(int[] features) {
        return dotProduct(features) > 0.0;
    }

    public void update(int[] features, boolean pass) {
        int d = pass ? 1 : -1;
        float dotp = dotProduct(features);

        for (int i = 0; i < len; i++) {
            weights[i] += rate * features[i] * (d - dotp);
        }

        bias += rate * (d - dotp);
    }
}
